// Copyright (c) 2021 All Right Reserved, B-com http://www.b-com.com/
//
// This file is subject to the B<>Com License.
// All other rights reserved.
//
// THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
// KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
// PARTICULAR PURPOSE.

syntax = "proto3";

package bcom.solar.cloud.rpc;

import "google/protobuf/empty.proto";

service SolARCloudProxy
{
    rpc Ping(google.protobuf.Empty) returns (google.protobuf.Empty);
    rpc SendMessage(Message) returns (google.protobuf.Empty);
    rpc SendFrames(Frames) returns (Pose);
    rpc SendFrame(Frame) returns (Pose);
    rpc StreamFrames(stream Frames) returns (Pose);
}

message Message
{
    string message = 1;
}

enum ImageLayout
{
    RGB_24 = 0;
    GREY_8 = 1;
    GREY_16 = 2;
}

message Image
{
    int32 width = 1;
    int32 height = 2;
    bytes data = 3;
    ImageLayout layout = 4;
}

message Frames
{
    repeated Frame frames = 1;
    fixed64 index = 2; // debug: server-side check of lost packet
}

message Frame
{
    int32 sensor_id = 1;
    Image image = 2;
    double timestamp = 3;
    Matrix4x4 pose = 4;
}

message Pose
{
    optional Matrix4x4 mat = 1;
}

message Matrix4x4
{
    float m11 = 1;
    float m12 = 2;
    float m13 = 3;
    float m14 = 4;
    float m21 = 5;
    float m22 = 6;
    float m23 = 7;
    float m24 = 8;
    float m31 = 9;
    float m32 = 10;
    float m33 = 11;
    float m34 = 12;
    float m41 = 13;
    float m42 = 14;
    float m43 = 15;
    float m44 = 16;
}
